library/monad/state.lisp

```diff
@@ -43,7 +45,27 @@ Represented as a closure from initial state to updated state and value."
     (match sc
       ((ST fstate)
        fstate)))
-  
+
+  ;;
+  ;; MonadState Typeclass
+  ;;
+
+  (define-class (Monad :m => MonadState :state :m (:m -> :state))
```

Comment:
It seems like this has some issues with accessor functions, which is causing the current failure in the brainfold example program.

This code:
```lisp
  (declare move-right (Unit -> (state:ST BF-State Unit)))
  (define (move-right)
    "Moves the pointer one bf-cell to the right."
    (do
     (bfs <- state:get)
     (pure (cell:increment! (.pointer bfs)))
      (state:put bfs)))
```
is throwing this error:
```
error: Ambiguous accessor
  --> <macroexpansion>:8:30
   |
 8 |        (PURE (CELL:INCREMENT! (.POINTER BFS)))
   |                                ^^^^^^^^ accessor is ambiguous
```
I'll dig into the accessor functions that the compiler generates and see if there's some way the type inference can be improved with the typeclass.